reconcile/deadmanssnitch.py

```diff
@@ -0,0 +1,226 @@
+            if not cluster.enable_dead_mans_snitch
+        ]
+        create_snitches = [
+            DiffData(cluster_name=cluster, action=Action.create_snitch)
+            # add cluster to create set only in case they do not exist in current state
+            for cluster in desired_cluster_enabled_true - current_state.keys()
+        ]
+        delete_snitches = [
+            DiffData(
+                cluster_name=cluster,
+                action=Action.delete_snitch,
+                data=current_state[cluster].token,
+            )
+            # add to delete set only if they exists in current state and flag is set to false
+            for cluster in desired_cluster_enabled_false & current_state.keys()
+        ]
+        update_vault = [
+            DiffData(
+                cluster_name=cluster,
+                action=Action.update_vault,
+                data=current_state[cluster].check_in_url,
+            )
+            for cluster in desired_cluster_enabled_true & current_state.keys()
+            # add to update set only if they exists, flag is set to true but vault and snitch url mismatch
+            if current_state[cluster].vault_data
+            and current_state[cluster].vault_data != current_state[cluster].check_in_url
+        ]
+        return create_snitches + delete_snitches + update_vault
+
+    def apply_diffs(
+        self, dry_run: bool, diffs: list[DiffData], diff_handler: DiffHandler
+    ) -> None:
+        if len(diffs) == 0:
+            return
+        logging.info(diff_handler.summarize(diffs=diffs))
+        if dry_run:
+            return
+        for diff in diffs:
+            try:
+                diff_handler.apply_diff(diff)
+            except Exception as e:
+                logging.error(str(e))
+                continue
+
+    def run(self, dry_run: bool) -> None:
+        # Initialize deadmanssnitch_api
+        settings = get_deadmanssnitch_settings()
+        token = self.secret_reader.read({
+            "path": settings.token_creds.path,
+            "field": settings.token_creds.field,
+        })
+        deadmanssnitch_api = DeadMansSnitchApi(token=token)
+        vault_client = cast(_VaultClient, VaultClient())
+        diff_handler = DiffHandler(deadmanssnitch_api, settings, vault_client)
+        # desired state - filter cluster having enableDeadMansSnitch field
+        clusters = [
+            cluster
+            for cluster in get_clusters_with_dms()
+            if cluster.enable_dead_mans_snitch is not None
+        ]
+        # current state - get snitches for tag app-sre
+        current_state = self.get_current_state(
+            deadmanssnitch_api, clusters, settings.snitches_path
+        )
+        if len(current_state) > 0 and len(clusters) > 0:
+            diff = self.get_diff(current_state=current_state, desired_state=clusters)
+            self.apply_diffs(dry_run, diff, diff_handler)
+        # close session
+        deadmanssnitch_api.close_session()
```

Comment:
`close_session` will be skipped when any exceptions happen before it, use `with` as context manager for `deadmanssnitch_api` to ensure it will be closed.